### PR TITLE
Enumerate mempool errors

### LINKED ISSUE #2590: Design: add enum variants for each transaction mempool error
## Motivation

We want to enumerate all of the possible reasons for ejecting transactions from the mempool or reasons for not adding a particular transaction for the mempool.

### PR DESCRIPTION
**Pull request by upbqdn**
## Motivation

Zebra needs an enum of errors for managing transactions in the mempool.

## Solution

I simply tried to enumerate the error cases.

## Review

@dconnolly, @teor2345.

## Follow Up Work

I presume many more cases will be added along the way.

### CODE REVIEW

File: zebrad/src/components/mempool/error.rs
Comment (oxarbitrage) [inline zebrad/src/components/mempool/error.rs]:
```suggestion
#[allow(dead_code)]
pub enum TransactionInsertError {
```

Should remove the never constructed warnings.

File: zebrad/src/components/mempool/error.rs
Comment (oxarbitrage) [inline zebrad/src/components/mempool/error.rs]:
```suggestion
#[allow(dead_code)]
pub enum TransactionQueryError {
```

**Review by oxarbitrage — COMMENTED** [review]
Maybe use `#[allow(dead_code)]` by now to remove the warnings until this stuff is used somewhere.

**Comment by teor2345** [comment]
I'm wondering how we want to handle transactions that get dropped after they are inserted into the mempool.
(For example, transactions that are too old, or a mempool that is too big.)

There's some overlap with `TransactionInsertError` variants, so we might want to work that out as part of this PR.
(For example, if a block containing a transaction is committed after the transaction enters the mempool, we want to drop that transaction.)

File: zebrad/src/components/mempool/error.rs
Comment (teor2345) [inline zebrad/src/components/mempool/error.rs]: Optional: We might want to be able to copy or clone these errors.
```suggestion
#[derive(Error, Copy, Clone, Debug, PartialEq)]
```

**Comment by mpguerra** [comment]
@ZcashFoundation/zebra-team where are we with this one? Is it enough as a start to merge for now or do we want any more changes to it?

File: zebrad/src/components/mempool/error.rs
Comment (oxarbitrage) [inline zebrad/src/components/mempool/error.rs]: Can the mempool update/delete transactions ?

If so, are we going to add possible errors into this enum ?

File: zebrad/src/components/mempool/error.rs
Comment (oxarbitrage) [inline zebrad/src/components/mempool/error.rs]:
```suggestion
    #[error("transaction is committed in block {0:?}")]
```

File: zebrad/src/components/mempool/error.rs
Comment (oxarbitrage) [reply zebrad/src/components/mempool/error.rs]: Definitely the mempool will drop transactions. If we drop by hash then it can fail if it is not found, if we just delete the first one the only failure will be if the mempool is empty.

Comment (oxarbitrage) [reply zebrad/src/components/mempool/error.rs]: I am just writing ideas that are coming to my mind for discussion, please feel free to ignore them.

What about a nested error structure as:

```rust
enum MempoolError {
    Insert,
    Drop,
    Query,
};
enum Insert {
   AlreadyExist,
   Invalid,
   ...
}
```

So, when we use them we can do stuff like `return Err(MempoolError::Insert::AlreadyExist)`.

File: zebrad/src/components/mempool/error.rs
Comment (dconnolly) [inline zebrad/src/components/mempool/error.rs]: I wonder if this is an error. This can definitely happen, but it seems to be part of normal behavior, vs a failure, so in such a case the query would return 'false' or empty set when looking for a transaction in the mempool.

@teor2345 @oxarbitrage what do you think?

File: zebrad/src/components/mempool/error.rs
Comment (teor2345) [reply zebrad/src/components/mempool/error.rs]: > What about a nested error structure

Rust enums don't quite work like this, so we usually do:

```
enum MempoolError {
    Insert(MempoolInsertError),
    Drop(MempoolDropError),
    Query(MempoolQueryError),
};
enum MempoolInsertError {
   AlreadyExist,
   Invalid,
   ...
}
impl From<MempoolInsertError> for MempoolError { ... }
```

And end up with code like `return Err(MempoolInsertError::AlreadyExist.into()).`

I think this change is optional.

File: zebrad/src/components/mempool/error.rs
Comment (teor2345) [reply zebrad/src/components/mempool/error.rs]: Maybe Blocker: This isn't an error in Zebra's current implementation:
https://github.com/ZcashFoundation/zebra/blob/bc4194fcb93e41b9482acb2a6626933c950d5695/zebra-network/src/protocol/internal/response.rs#L17-L26

The first table in this document says how Zebra responds to each inbound mempool message:
https://docs.google.com/document/d/1KgPl3r_Pp8AUM0KvDnBv7LK0vRMvxVjM9ooIwr2mB7U/edit#

File: zebrad/src/components/mempool/error.rs
Comment (teor2345) [inline zebrad/src/components/mempool/error.rs]: Optional naming style: In Rust, we try to avoid repeating words in the enum type and variant names.

For example: `TransactionInsertError::Invalid`, not `TransactionInsertError::TransactionInvalid`.

File: zebrad/src/components/mempool/error.rs
Comment (teor2345) [reply zebrad/src/components/mempool/error.rs]: Maybe Blocker

> Can the mempool update/delete transactions ?
> 
> If so, are we going to add possible errors into this enum ?

I also asked a similar question at:
https://github.com/ZcashFoundation/zebra/pull/2615#issuecomment-898055391

> I'm wondering how we want to handle transactions that get dropped after they are inserted into the mempool.
> (For example, transactions that are too old, or a mempool that is too big.)
> 
> There's some overlap with TransactionInsertError variants, so we might want to work that out as part of this PR.
> (For example, if a block containing a transaction is committed after the transaction enters the mempool, we want to drop that transaction.)

Since the mempool service is blocking on this PR, here's what I suggest:
1. Use a single `MempoolError` type in this PR
2. Open a follow-up ticket to split the errors into categories, once we know what all the errors are

**Review by teor2345 — COMMENTED** [review]
Here are some suggestions for improvements.

2 of them might be blockers, because they could make the error type confusing or hard to use.

The rest are optional.

**Comment by teor2345** [comment]
> @ZcashFoundation/zebra-team where are we with this one? Is it enough as a start to merge for now or do we want any more changes to it?

I think these two changes might be blockers for this PR:
https://github.com/ZcashFoundation/zebra/pull/2615#issuecomment-898055391
https://github.com/ZcashFoundation/zebra/pull/2615#discussion_r694421383

But I'm happy to have @dconnolly do the review and merge here, because they are waiting on this PR.

File: zebrad/src/components/mempool/error.rs
Comment (upbqdn) [reply zebrad/src/components/mempool/error.rs]: I removed `TransactionQueryError` then.

File: zebrad/src/components/mempool/error.rs
Comment (teor2345) [reply zebrad/src/components/mempool/error.rs]: Oops, sorry, this didn't work, because the error types we're wrapping don't implement `Copy` or `Clone`.

I'll see if I can fix that now.

File: zebrad/src/components/mempool/error.rs
Comment (teor2345) [inline zebrad/src/components/mempool/error.rs]: Oops, this was my mistake.

`Clone` should work after PR #2664 merges, and `main` is merged into this PR.

Or feel free to revert the entire change, and do it in a later PR.

```suggestion
#[derive(Error, Clone, Debug, PartialEq)]
```

File: zebrad/src/components/mempool/error.rs
Comment (teor2345) [reply zebrad/src/components/mempool/error.rs]: I made a comment with a suggested fix.

**Comment by upbqdn** [comment]
Yea, I'm sorry, I didn't know how to respond to https://github.com/ZcashFoundation/zebra/pull/2615#issuecomment-898055391.

**Comment by teor2345** [comment]
> Yea, I'm sorry, I didn't know how to respond to [#2615 (comment)](https://github.com/ZcashFoundation/zebra/pull/2615#issuecomment-898055391).

No worries at all.

I'm happy to make specific suggestions if they're needed - just let me know.
But I want to leave people the freedom to come up with their own way forward first.

**Comment by upbqdn** [comment]
I used a single `MempoolError` type, and I opened https://github.com/ZcashFoundation/zebra/issues/2668 to split the errors into categories as suggested in https://github.com/ZcashFoundation/zebra/pull/2615#discussion_r694423838.

**Review by oxarbitrage — APPROVED** [review]
Looks good.

**Review by dconnolly — APPROVED** [review]

**Comment by dconnolly** [comment]
I have some ideas on this as part of the Mempool Storage tracking, but I don't think this PR is blocked on that.

**Pull request merged**
